### ISSUE TITLE
Single byte (or 2) prevent finding a word

### ISSUE THREAD

**Issue by cocoon** [dnGrep/dnGrep#1164]
I have a problem to find a word in special java class files.

I found out, if I edit 1 Byte to 00 or I found even another sequense of 2 byte later that can make it findable again. 

Any chance to make it working with the original example?

I prepared 3 example files in this ZIP:
[example.zip](https://github.com/user-attachments/files/16439499/example.zip)

the original head part where it cannot find the word "TEST":
EXAMPLE_Cannot_Find_TEST.txt

and the two modified one where I made the mentioned changes:

EXAMPLE_Can_Find_TEST-01.txt
changed the very start of the file 
from: CA FE BA BE
To: CA FE BA 00

EXAMPLE_Can_Find_TEST-02.txt
changed at offset 3F5
from: 42 C8
to: 00 00

**Comment by doug24** [comment]
This is a problem with the file encoding - java class files are binary, not text files. dnGrep runs an encoding detector on all text files, but not on binary files. The encoding detection logic used by dnGrep doesn't work on binary files.  

Your EXAMPLE_Cannot_Find_TEST.txt file is binary but fails the binary test in dnGrep (it looks for 4 consecutive nulls in the first 1kb), so it thinks it is plain text. The two files you modified are detected as binary.

For text files, the encoding detector runs.  On EXAMPLE_Cannot_Find_TEST.txt it is detected as Unicode (UTF-16, little endian).  Your "TEST TEST TEST" does not get converted to those characters using 16-bit encoding.

For binary files, the encoding used it just the system default, or whatever you select for an encoding in the encoding selector. On my system the default is UTF-8, but that will vary, but is usually some 8-bit encoding.

If you set the encoding selector in dnGrep to UTF-8 (or any other 8-bit encoding), dnGrep will use that encoding and it will find "TEST" in all three files.

<img src="https://github.com/user-attachments/assets/1f23d4a8-76e7-4ce5-9f39-c638677dbb89" width="400"/>

**Comment by cocoon** [comment]
Ah wonderful thank you so much for explaining!
And yes I can confirm if I manually set it to UTF-8 I can find the text in the original class files in jar archives.

Maybe adding an advanced option to do a long / deep search which is trying every different encoding would be helpful?
Of course this would take longer, but would find more without the need to manually try switching encodings.

**Comment by doug24** [comment]
I have 116 different encodings on my version of Windows. It may depend on the region settings, but that is probably typical.  I don't think many people would want to reopen each file and search it again 116 times (worst case if the file has no matches).

It would be great if there was a more accurate way to determine if a file is binary or not so you wouldn't get the false negative. But I don't know of a better way to check.

The other problem searching binary files is that I have seen binary files with embedded text where the text has different encodings in the same file.  Some was 7/8-bit encoding, and some was 16-bit encoding.  If trying all the encodings and stopping at the first encoding that generates a match, you might miss other matches that use a different encoding.

Treating a binary file as a text file is essentially a hack; it might work, but it's not guaranteed.

**Comment by cocoon** [comment]
"Stop after first match" is already a search option, so by default it should really scan for everything that makes sense.
If the intent is to really FIND something, I would of courser accept that it may take much longer.

For me it might have been enough to use the byte search (a quick / automatic converter / input from text to byte sequence would be nice)
But currently the Byte search seems not to search in archives, that would be good to have there, too.

**Comment by doug24** [comment]
You want to see all of them: Since you are searching for ASCII characters, about 70 of the 116 possible encodings will correctly decode TEST because they all use the same byte values for 0-9, a-z, and A-Z.  If the search string used non-ASCII characters, the 70 may be reduced greatly. But now we have 70 sets of results with matches for just one file.  Does dnGrep show all 70 sets for each file so that they results are shown in the respective encoding?  Or should they be collapsed into one for all 7/8-bit encodings?  If so, which encoding should be used?  Maybe the user isn't interested in just the search string, but also the data around the search string - the encoding used could make a big difference.

I think if the user knows anything about the file/file source they can make an educated guess of a couple possible encodings.  If it is likely a Western European language, it is not likely to be Chinese, Japanese, or Cyrillic. 

I have thought about getting the encoding from the search text, but getting the encoding from just 4 characters is indeterminate - it could be any one of 70.  It will just end up being the system default encoding.

Yes, you can use a byte search.  But when you enter your search values, it will be case sensitive, and it will assume single byte or multibyte encoding.  To search for TEST with single byte encoding you need `54 45 53 54`, and for multibyte (Unicode) you need `54 00 45 00 53 00 54 00`.  

Also note the text area of the byte search results once again requires an encoding to convert the bytes to characters.  And if we don't know the correct encoding, the preview will not show what you expect.  Give it a try.

You are correct, the byte search does not work in archives.  I'll look into that sometime.

**Comment by cocoon** [comment]
Yes I understand the encoding problem, that is why I think this is a perfect work for a program, because it is much better than to manually switch through all encodings manually and search again.

The byte case was just my special case as I already found my sequence and so knew the byte sequence.

Just to explain my current use case:

I had to use a program that was throwing a special error message and I needed to find out where it comes from to debug it.
I did not know anything about how the program stores it, resources, clear text, xml ... could have been anything, that is why I wanted to deep search for it.

Previously I used grepWin, but needed something that can search inside archives (i.e. .jar) and so found dnGrep, and I like it. 👍
